sqld/src/rpc/wal_log.rs

```diff
@@ -74,8 +85,32 @@ impl WalLog for WalLogService {
         &self,
         req: tonic::Request<LogOffset>,
     ) -> Result<tonic::Response<Self::LogEntriesStream>, Status> {
+        let replica_addr = req
+            .remote_addr()
+            .ok_or(Status::internal("No remote RPC address"))?;
+        {
+            let guard = self.replicas_with_hello.read().await;
+            if !guard.contains(&replica_addr) {
+                return Err(Status::failed_precondition("Hello message missing"));
+            }
+        }
         let start_offset = req.into_inner().start_offset;
         let stream = self.stream_pages(start_offset as _);
         Ok(tonic::Response::new(stream))
     }
+    async fn hello(
+        &self,
+        req: tonic::Request<HelloRequest>,
+    ) -> Result<tonic::Response<HelloResponse>, Status> {
+        let replica_addr = req
+            .remote_addr()
+            .ok_or(Status::internal("No remote RPC address"))?;
+        {
+            let mut guard = self.replicas_with_hello.write().await;
+            if !guard.contains(&replica_addr) {
+                guard.push(replica_addr);
+            }
```

Comment:
let's just use an hashset to make this operation idempotent

Reply:
hashset is an overkill when we expect << 100 replicas

Reply:
it's just more ergonomic, and expresses the intent better